air/src/execution_step/instructions/ap_map.rs

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2023 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use super::ap::utils::generate_map_value_descriptor;
+use super::ExecutionCtx;
+use super::ExecutionResult;
+use super::TraceHandler;
+use crate::execution_step::execution_context::errors::StreamMapError::FloatMapKeyIsUnsupported;
+use crate::execution_step::execution_context::stream_map_key::StreamMapKey;
+use crate::execution_step::execution_context::StreamMapValueDescriptor;
+use crate::execution_step::instructions::ap::apply_to_arguments::apply_to_arg;
+use crate::execution_step::resolver::Resolvable;
+use crate::execution_step::ValueAggregate;
+use crate::log_instruction;
+use crate::trace_to_exec_err;
+use crate::CatchableError;
+use crate::ExecutionError;
+
+use air_interpreter_data::GenerationIdx;
+use air_parser::ast::ApMap;
+use air_parser::ast::ApMapKey;
+use air_parser::ast::Number;
+use air_parser::ast::StreamMap;
+use air_trace_handler::merger::MergerApResult;
+
+impl<'i> super::ExecutableInstruction<'i> for ApMap<'i> {
+    #[tracing::instrument(level = "debug", skip(exec_ctx, trace_ctx))]
+    fn execute(&self, exec_ctx: &mut ExecutionCtx<'i>, trace_ctx: &mut TraceHandler) -> ExecutionResult<()> {
+        log_instruction!(call, exec_ctx, trace_ctx);
+        // this applying should be at the very beginning of this function,
+        // because it's necessary to check argument lambda, for more details see
+        // https://github.com/fluencelabs/aquavm/issues/216
+        let result = apply_to_arg(&self.value, exec_ctx, trace_ctx, true)?;
+
+        let merger_ap_result = to_merger_ap_map_result(&self, trace_ctx)?;
+        let generation = populate_context(&self.key, &self.map, &merger_ap_result, result, exec_ctx)?;
+        maybe_update_trace(generation, trace_ctx);
+
+        Ok(())
+    }
+}
+
+fn to_merger_ap_map_result(instr: &impl ToString, trace_ctx: &mut TraceHandler) -> ExecutionResult<MergerApResult> {
+    let merger_ap_result = trace_to_exec_err!(trace_ctx.meet_ap_start(), instr)?;
+    Ok(merger_ap_result)
+}
+
+fn populate_context<'ctx>(
+    key: &ApMapKey<'ctx>,
```

Comment:
Argument order seems to be arbitrary.

Reply:
Elaborate on this.

Reply:
The key is apart from value, for example.

Reply:
I c. A subjective point of view so I will choose to pass on this.

Reply:
I basically agree with @monoid and propose refactoring this code close to the following way:
```rust
fn execute(&self, exec_ctx: &mut ExecutionCtx<'i>, trace_ctx: &mut TraceHandler) -> ExecutionResult<()> {
     ...
     let resolved_key = resolve(...);
     let generation = populate_context(resolved_key, value, ...)?;
     maybe_update_trace(generation, trace_ctx);

     Ok(())
}

fn populate_context<'ctx>(
    key: &StreamMapKey<'ctx>,
    value: ValueAggregate,
    ap_map_result: &StreamMap<'ctx>,
    merger_ap_result: &MergerApResult,
    exec_ctx: &mut ExecutionCtx<'ctx>,
) -> ExecutionResult<GenerationIdx> {
    let value_descriptor = generate_map_value_descriptor(result, ap_map_result, merger_ap_result);
    exec_ctx
          .stream_maps
          .add_stream_map_value(key, value_descriptor)
}

fn resolve_key(
    key: &impl Resolvable,
    exec_ctx: &mut ExecutionCtx<'_>,
    value_descriptor: StreamMapValueDescriptor<'_>,
) -> Result<StreamMapKey, ExecutionError> {
...
}
```